javascript/ql/src/semmle/javascript/dataflow/Configuration.qll

```diff
@@ -792,15 +786,108 @@ private predicate isSink(DataFlow::Node nd, DataFlow::Configuration cfg, FlowLab
   cfg.isSink(nd, lbl)
 }
 
+/**
+ * Holds if there exists a load-step from `pred` to `succ` under configuration `cfg`,
+ * and the forwards exploratory flow has found a relevant store-step with the same property as the load-step.
+ */
+private predicate exploratoryLoadStep(
+  DataFlow::Node pred, DataFlow::Node succ, DataFlow::Configuration cfg
+) {
+  exists(string prop | prop = getAForwardRelevantLoadProperty(cfg) |
+    isAdditionalLoadStep(pred, succ, prop, cfg)
+    or
+    basicLoadStep(pred, succ, prop)
+  )
+}
+
+/**
+ * Gets a property where the forwards exploratory flow has found a relevant store-step with that property.
+ * The property is therefore relevant for load-steps in the forward exploratory flow.
+ *
+ * This private predicate is only used in `exploratoryLoadStep`, and exists as a separate predicate to give the compiler a hint about join-ordering.
+ */
+private string getAForwardRelevantLoadProperty(DataFlow::Configuration cfg) {
+  exists(DataFlow::Node previous | isRelevantForward(previous, cfg) |
+    basicStoreStep(previous, _, result) or
+    isAdditionalStoreStep(previous, _, result, cfg)
+  )
+  or
+  result = getAPropertyUsedInLoadStore(cfg)
+}
+
+/**
+ * Gets a property that is used in an `additionalLoadStoreStep` where the loaded and stored property are not the same.
+ *
+ * The properties from this predicate are used as a white-list of properties for load/store steps that should always be considered in the exploratory flow.
+ */
+private string getAPropertyUsedInLoadStore(DataFlow::Configuration cfg) {
+  exists(string loadProp | not loadProp = result |
```

Comment:
As a matter of readability, I'd suggest putting `and loadProp != result` after the condition that binds `loadProp` instead of opening with `not loadProp = result`.

Reply:
I came up with another refactorization where I combined the two cases in the predicate.  
And I applied your suggestion to that refactorization.

Reply:
Thanks, much better. I did a quick tuple count diff based on the Xss query and the change seems safe to land without (yet) another round of evaluations.